plutus-ir/src/Language/PlutusIR/Analysis/Dependencies.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE ConstraintKinds  #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE LambdaCase       #-}
+-- | Functions for computing the dependency graph of variables within a term or type.
+module Language.PlutusIR.Analysis.Dependencies (Node (..), DepGraph, runTermDeps, runTypeDeps) where
+
+import qualified Language.PlutusCore               as PLC
+import qualified Language.PlutusCore.Name          as PLC
+
+import           Language.PlutusIR
+import qualified Language.PlutusIR.Analysis.Usages as Usages
+
+import           Control.Monad.Reader
+
+import qualified Algebra.Graph.Class               as G
+import           Data.Coerce
+import qualified Data.Set                          as Set
+import           Lens.Micro
+
+-- | A node in a dependency graph. Either a specific 'PLC.Unique', or a specific
+-- node indicating the root of the graph. We need the root node because when computing the
+-- dependency graph of, say, a term, there will not be a binding for the term itself which
+-- we can use to represent it in the graph.
+data Node = Unique PLC.Unique | Root deriving (Show, Eq, Ord)
+
+-- | A constraint requiring @g@ to be a 'G.Graph' (so we can compute e.g. a @Relation@ from it), whose
+-- vertices are 'Node's.
+type DepGraph g = (G.Graph g, (G.Vertex g)~Node)
+
+-- | Compute the dependency graph of a 'Term'. The 'Root' node will correspond to the term itself.
+runTermDeps
+    :: (DepGraph g, PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+    => Term tyname name a
+    -> g
+runTermDeps t = runReader (termDeps t) Root
+
+-- | Compute the dependency graph of a 'Type'. The 'Root' node will correspond to the type itself.
+runTypeDeps
+    :: (DepGraph g, PLC.HasUnique (tyname a) PLC.TypeUnique)
+    => Type tyname a
+    -> g
+runTypeDeps t = runReader (typeDeps t) Root
+
+-- | Record some dependencies on the current node.
+recordDeps
+    :: (DepGraph g, MonadReader Node m)
+    => [PLC.Unique]
+    -> m g
+recordDeps us = do
+    current <- ask
+    pure $ G.connect (G.vertices [current]) (G.vertices (fmap Unique us))
+
+-- | Process the given action with the given name as the current node.
+withCurrent
+    :: (MonadReader Node m, PLC.HasUnique n u)
+    => n
+    -> m g
+    -> m g
+withCurrent n = local (const $ Unique $ coerce $ n ^. PLC.unique)
```

Comment:
`const $ coerce $ n ^. PLC.unique` does not work? Also I'd prefer `n ^. PLC.unique . coerced`, but that's a matter of taste, of course.

Reply:
This is my fault for reusing the type name - `Unique` here is not `PLC.Unique`, let me rename that. The other suggestion is nice, though.